### PR TITLE
feat(ai): Experimental AI suggest feature for event page

### PR DESCRIPTION
**Pull request by mitsuhiko**
This adds an experimental popover onto the event page that uses ChatGPT to explain the error. This is not intended to be shipped, but just a small playground to see if something like this could be useful. Unfortunately this pulls in some dependencies which are not yet on our package index. Unclear if we want to add the openai packages there for this experiment.

image

<details><summary><strong>More Screenshots</strong></summary>

image

</details>

How to experiment with this:

* The prompt used is located in `event_ai_suggest.py`, in the `PROMPT` variable
* To use this, you need to inject an API key into your `sentry.conf.py` (`OPENAI_API_KEY = "sk-blablabla"`)
* You need to enable the `organizations:ai-suggest` feature (eg: `SENTRY_FEATURES["organizations:ai-suggest"] = True` in `sentry.conf.py`)

## Todo

* [x] basic legal review
  * [ ] legal review of OpenAI DPA
  * [ ] clearly reference this as a "no-charge experimental feature"
  * [ ] for invoiced (touch sales) or customers that signed in the DPA in app, block access to the feature until a DPA addendum is signed
  * [ ] for all other customers bring up a dialog once, that using the feature sends data to OpenAI
* [ ] security review of OpenAI
* [ ] design review of modal solution

### CODE REVIEW

File: src/sentry/api/endpoints/event_ai_suggest.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        event = eventstore.get_event_by_id(project.id, event_id)
+        if event is None:
+            raise ResourceDoesNotExist
+
+        if not features.has("organizations:ai-suggest", project.organization, actor=request.user):
+            raise ResourceDoesNotExist
+
+        # Cache the suggestion for a certain amount by primary hash, so even when new events
+        # come into the group, we are sharing the same response.
+        cache_key = "ai:" + event.get_primary_hash()
```
Comment (wedamija) [inline src/sentry/api/endpoints/event_ai_suggest.py:191]: Should we just use the `group_id` instead?

Does it also make sense to restrict this only to error issues?

File: src/sentry/api/endpoints/event_ai_suggest.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+FUN_PROMPT_CHOICES = [
+    "[haiku about the error]",
+    "[hip hop rhyme about the error]",
+    "[4 line rhyme about the error]",
+    "[2 stanza rhyme about the error]",
+    "[anti joke about the error]",
+]
+
+PROMPT = """\
+You are an assistent that analyses software errors, describing the problem with the follwing rules:
```
Comment (wedamija) [inline src/sentry/api/endpoints/event_ai_suggest.py:35]: Nit: assistant, following

Not sure if typos affect the prompt but probably worth fixing.

File: src/sentry/api/endpoints/event_ai_suggest.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+from sentry.utils.cache import cache
+
+logger = logging.getLogger(__name__)
+
+
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+# XXX: this is pretty ugly
+openai.api_key = settings.OPENAI_API_KEY
```
Comment (wedamija) [inline src/sentry/api/endpoints/event_ai_suggest.py:24]: Probably worth just moving this to server startup or something like that

File: src/sentry/api/endpoints/event_ai_suggest.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+        This endpoint returns a JSON response that provides helpful suggestions about how to
+        understand or resolve an event.
+        """
+        event = eventstore.get_event_by_id(project.id, event_id)
+        if event is None:
+            raise ResourceDoesNotExist
+
+        if not features.has("organizations:ai-suggest", project.organization, actor=request.user):
+            raise ResourceDoesNotExist
```
Comment (wedamija) [inline src/sentry/api/endpoints/event_ai_suggest.py:187]: Should we also exit early here if no openai token is present?

File: src/sentry/api/endpoints/event_ai_suggest.py (hunk 1 above)
Comment (mitsuhiko) [reply src/sentry/api/endpoints/event_ai_suggest.py:191]: Unclear yet how this works with issue platform issues, want to actually discuss this with you. From what I can tell right now `get_event_by_id` from the event store returns an event with `group_id`, but our docs on that also say that we consider this deprecated which is why I did not want to build this into the cache.

I assume we're asking for the group ID as additional information, but I do not believe that the API actually needs that info for anything other than the cache key.

Comment (wedamija) [reply src/sentry/api/endpoints/event_ai_suggest.py:195]: Yeah, we need to clean this up a bit. Once performance issues have moved over to the issue platform, I can clean up a bunch of cruft here and make things clearer - at the moment we're in this split world where events can still have multiple groups associated with them, which is going away once perf issues are migrated. Once that's done, it'll be easier to clean up the rest of the interface here as well.

**Comment by priscilawebdev** [comment]
closing this PR as the implementation was done in the following ones:

- https://github.com/getsentry/sentry/pull/46375 (backend)
- https://github.com/getsentry/sentry/pull/46384 (frontend)

**Pull request closed without merging**